x/pairing/keeper/msg_server_relay_payment.go

```diff
@@ -48,91 +49,102 @@ func (k msgServer) RelayPayment(goCtx context.Context, msg *types.MsgRelayPaymen
-				badgeUsedCuTimerExpiryBlock := k.BadgeUsedCuExpiry(ctx, badgeData.Badge)
-				if badgeUsedCuTimerExpiryBlock <= uint64(ctx.BlockHeight()) {
-					return nil, utils.LavaFormatWarning("badge rejected", fmt.Errorf("badge used CU entry validity expired"),
-						utils.Attribute{Key: "badgeUsedCuTimerExpiryBlock", Value: badgeUsedCuTimerExpiryBlock},
-						utils.Attribute{Key: "currentBlock", Value: uint64(ctx.BlockHeight())},
-					)
-				}
-				// timerKey = badgeUsedCuMapKey since all badgeUsedCuMapKey keys are unique - can be used to differentiate the timers
-				k.badgeTimerStore.AddTimerByBlockHeight(ctx, badgeUsedCuTimerExpiryBlock, badgeUsedCuKey, []byte{})
-
-				badgeUsedCuMapEntry = types.BadgeUsedCu{
-					BadgeUsedCuKey: badgeUsedCuKey,
-					UsedCu:         0,
-				}
-			}
-
-			// enforce badge CU overuse
-			if relay.CuSum+badgeUsedCuMapEntry.UsedCu > badgeData.Badge.CuAllocation {
-				return nil, utils.LavaFormatWarning("badge CU allocation exceeded", fmt.Errorf("could not update badge's used CU"),
-					utils.Attribute{Key: "relayCuSum", Value: relay.CuSum},
-					utils.Attribute{Key: "badgeCuLeft", Value: badgeData.Badge.CuAllocation - badgeUsedCuMapEntry.UsedCu},
-				)
+			newBadgeTimerExpiry, err = k.checkBadge(ctx, badgeData, clientAddr.String(), relay)
+			if err != nil {
+				utils.LavaFormatWarning("badge check failed", err)
+				rejectedCu += relay.CuSum
+				continue
 			}
 
-			badgeUsedCuMapEntry.UsedCu += relay.CuSum
-			k.SetBadgeUsedCu(ctx, badgeUsedCuMapEntry)
-
 			// badge is valid & CU enforced -> switch address to badge signer (developer key) and continue with payment
 			clientAddr = badgeData.BadgeSigner
 			badgeSig = badgeData.Badge.ProjectSig
 		}
 
 		providerAddr, err := sdk.AccAddressFromBech32(relay.Provider)
 		if err != nil {
-			return nil, utils.LavaFormatWarning("invalid provider address in relay msg", err,
+			utils.LavaFormatWarning("invalid provider address in relay msg", err,
 				utils.Attribute{Key: "provider", Value: relay.Provider},
 				utils.Attribute{Key: "creator", Value: msg.Creator},
 			)
+			rejectedCu += relay.CuSum
+			continue
 		}
 		if !providerAddr.Equals(creator) {
-			return nil, utils.LavaFormatWarning("invalid provider address in relay msg", fmt.Errorf("creator and signed provider mismatch"),
+			utils.LavaFormatWarning("invalid provider address in relay msg", fmt.Errorf("creator and signed provider mismatch"),
 				utils.Attribute{Key: "provider", Value: relay.Provider},
 				utils.Attribute{Key: "creator", Value: msg.Creator},
 			)
+			rejectedCu += relay.CuSum
+			continue
+		}
+
+		project, err := k.GetProjectData(ctx, clientAddr, relay.SpecId, uint64(relay.Epoch))
+		if err != nil {
+			utils.LavaFormatWarning("invalid project data", err)
+			rejectedCu += relay.CuSum
+			continue
+		}
+
+		epochStart, _, err := k.epochStorageKeeper.GetEpochStartForBlock(ctx, uint64(relay.Epoch))
```

Comment:
divide to 2 if cases
if err != nil {
}

 if k.IsDoubleSpend(ctx, relay.SpecId, epochStart, project.Index, providerAddr, strconv.FormatUint(relay.SessionId, 16)) {
...
}

Reply:
done. see https://github.com/lavanet/lava/commit/1090981f67cb42d98190a484b5bbfc0bd645c78e